plugin/pulpcore/plugin/repository.py

```diff
@@ -0,0 +1,171 @@
+        return self._model.number
+
+    @property
+    def content(self):
+        """
+        QuerySet of content on this repository version
+
+        Returns:
+            QuerySet: The Content objects that are related to this version.
+        """
+        return self._model.content
+
+    def add_content(self, content):
+        """
+        Add a content unit to the repository.
+
+        Args:
+           content (pulpcore.plugin.models.Content): a content model to add
+        """
+        assert not self._model.complete, _("Cannot update version {version} for repository {repo}."
+                                           " Complete Repository is immutable")\
+            .format(version=self._model.number, repo=self._model.repository.name)
+
+        with suppress(IntegrityError):
+            association = models.RepositoryContent(
+                repository=self._model.repository,
+                content=content,
+                version_added=self._model)
+            association.save()
+
+    def remove_content(self, content):
+        """
+        Remove content from the repository.
+
+        Args:
+            content (pulpcore.plugin.models.Content): A content model to remove
+        """
+        assert not self._model.complete, _("Cannot update version {version} for repository {repo}."
+                                           " Complete Repository is immutable")\
+            .format(version=self._model.number, repo=self._model.repository.name)
+
+        q_set = models.RepositoryContent.objects.filter(
+            repository=self._model.repository,
+            content=content,
+            version_removed=None)
+        q_set.update(version_removed=self._model)
+
+    def complete(self):
+        """
+        Save the repository version when all operations are done on it
+        """
+        with transaction.atomic():
+            self._model.complete = True
+            self._model.save()
+
+    def delete(self):
+        """
+        Deletes an incomplete Repository Version
+
+        This method deletes a RepositoryVersion only if its 'complete' property is False. All
+        RepositoryContent added in the deleted RepositoryVersion are deleted also. All
+        RepositoryContent removed in the deleted RepositoryVersion has the version_removed set to
+        None.
+        """
+        assert not self._model.complete, _("Version {version} for repository {repo} is complete.").\
+            format(version=self._model.number, repo=self._model.repository.name)
+        with transaction.atomic():
+            models.RepositoryContent.objects.filter(version_added=self._model).delete()
+            models.RepositoryContent.objects.filter(version_removed=self._model)\
+                .update(version_removed=None)
+            self._model.repository.last_version = self._model.number - 1
+            self._model.repository.save()
+            self._model.delete()
+            if self._created_resource is not None:
```

Comment:
Suggest `CreatedResource.objects.delete(content_object=self._model)` so you don't need to store `self._created_resource` as an attribute.  It would be cleaner.

Reply:
I am not sure if this is possible without updating CreatedResource.content_object to be a GenericRelation [0] 
Running `` models.CreatedResource.objects.filter(content_object=self._model).delete()`` generates the error message
``django.core.exceptions.FieldError: Field 'content_object' does not generate an automatic reverse relation and therefore cannot be used for reverse querying. If it is a GenericForeignKey, consider adding a GenericRelation.``

[0] https://stackoverflow.com/questions/36164654/how-to-traverse-a-genericforeignkey-in-django/36166644#36166644